### DISCUSSION TITLE
OnAfterConversionToEntity - Convert String to Int before return endpoints

### DISCUSSION THREAD

**Discussion by RecoX** [ravendb/ravendb#15200] [Q&A] [answered]
Currently, I am working on a migration from Raven 3.5 to Raven 5, the system I am maintaining relies heavily on the Ids as ints but Raven5 uses Ids as strings.

For example:
`Products/1` instead of `1`

In the beginning, I was doing an anti-pattern to make it work, in the return of some endpoint I just modify manually the ID like in the following example:

```
        [Route("GetCatering")]
        [ProducesResponseType(StatusCodes.Status200OK)]
        [ProducesResponseType(StatusCodes.Status404NotFound)]
        [ProducesResponseType(typeof(Menu), 200)]
        [HttpGet]
        public IActionResult GetCatering()
        {
            var menu =  _sessionOnlineOrdering.Query<Menu>()
                                               .Include(x => x.Combos)
                                               .Include(x => x.Products)
                                               .Include(x => x.ProductsDetails.Select(p => p.ModifiersIds))
                                               .Where(x => x.IsCateringMenu)
                                               .FirstOrDefault();

            if (menu == null) {
                return NotFound("There is no menu for catering");
            }

            menu.Id = menu.Id.Replace("menus/", "");

            _menusHelper.GetProductsFromMenu(menu, _sessionOnlineOrdering);
            _menusHelper.GetCombosFromMenu(menu, _sessionOnlineOrdering);

            return Ok(menu);
        }
```

it is very dirty but it helped me to continue doing my work
I need to solve this properly, using a transformer or a parser or something.

Talking with @ayende , he suggests using an in-built functionality called store.Conventions like in the following picture:

![image](https://user-images.githubusercontent.com/5874806/199152915-b844038a-11c5-4e8f-926a-3ba5f0028e80.png)

The thing is, in my Raven version 5.3.102 (I am keen to update to the latest one always now) I can't find that function but I found another thing `store.OnAfterConversionToEntity` which looks very similar to that.

I am trying to use it to do the conversion of the strings to ints in the returns of every endpoint, doing something like this throws an error as args is just read-only
```

            store.OnAfterConversionToEntity += (sender, args) => {
                args.Id = 123;
                // return 1234;
                // args.Id.ToString();
            };
```

Can someone give me some help to solve my problem?

Thanks

**Comment by JoshuaLevett** [comment]
> but Raven5 uses Ids as int

This isn't true raven is perfectly happy with a document with the Id `Foo/Bar`

`Products/1` doesn't look like a normal Raven Id. I'd expect it to include the node in the id, e.g. `Products/1-A`.
Raven lets you control what your Id conventions are, it's possible that might be all you need to look at.
[https://ravendb.net/docs/article-page/5.1/Csharp/client-api/configuration/identifier-generation/global](https://ravendb.net/docs/article-page/5.1/Csharp/client-api/configuration/identifier-generation/global)

As far as a solution to what it looks like you are doing you could try adding a property like the following to `Menu`
`public string StrippedId { get => Id.StripCollectionName(); }`

Where `StripCollectionName()` is a string extension like the below.
```
public static string StripCollectionName(this string id)
        {
            if (string.IsNullOrEmpty(id)) return string.Empty;
            var slashIndex = id.IndexOf("/");
            if (slashIndex < 0)
                return id;
            }
            return id.Substring(slashIndex + 1);
        }
```

This property will even be stored in the document and can be queried against / fetched in `.Select`.
You could even change it to return an int

Hopefully one of those is practical for you.

**Reply by RecoX** [reply]
> > but Raven5 uses Ids as int
> 
> This isn't true raven is perfectly happy with a document with the Id `Foo/Bar`
> 

Sorry. a typo Raven5 uses Ids as strings I wanted to say


> `Products/1` doesn't look like a normal Raven Id. I'd expect it to include the node in the id, e.g. `Products/1-A`. Raven lets you control what your Id conventions are, it's possible that might be all you need to look at. https://ravendb.net/docs/article-page/5.1/Csharp/client-api/configuration/identifier-generation/global
> 

For me is still using Entity/1 as I can show you in the image
![image](https://user-images.githubusercontent.com/5874806/199346726-877b4790-947c-4d7a-a052-b9a3c7ec0237.png)


I am replicating from 3.5 to 5, but also I created entities directly in Raven 5 and still maintain the Entity/# which is very ok for me, I do not need the -NodeName (I understand is good for decentralized systems, but our system is very basic and we do not need that feature), indeed with just Ids as integers we are fine.

> As far as a solution to what it looks like you are doing you could try adding a property like the following to `Menu` `public string StrippedId { get => Id.StripCollectionName(); }`
> 
> Where `StripCollectionName()` is a string extension like the one below.
> 
> ```
> public static string StripCollectionName(this string id)
>         {
>             if (string.IsNullOrEmpty(id)) return string.Empty;
>             var slashIndex = id.IndexOf("/");
>             if (slashIndex < 0)
>                 return id;
>             }
>             return id.Substring(slashIndex + 1);
>         }
> ```
> 
> This property will even be stored in the document and can be queried against / fetched in `.Select`. You could even change it to return an int
> 
> Hopefully one of those is practical for you.

I would love to avoid that solution so my Class don't get polluted with extra behaviour for this, my intention is to somehow use something as a middleware like AfterConversionToEntity and do something like this in just one place to change the Ids when I am returning a response, I think that's my best bet to avoid change deep things in Raven 5 and make my consumers work as they expect.

```
AfterConversionToEntity(entity) {
            if (string.IsNullOrEmpty(entity.id)) return string.Empty;
            var slashIndex = entity.id.IndexOf("/");
            if (slashIndex < 0)
                return entity.id;
            }
            return entity.id.Substring(slashIndex + 1);
}
```

**Reply by georgiosd** [reply]
I don't think that's a good idea - removing the node name can result in conflicts. Are you deploying this to a cluster or a single node instance?

**Reply by RecoX** [reply]
I did not do that, Raven did it for me because that is deployed in a single instance as is just for Dev and Production is deployed on a cluster, anyways I do not care about it, if is there is all good, if it is not there I really don't care, I just care about the output of the endpoints with the OnAfterConversionToEntity event or some alternative to solve my problem.

**Reply by garayx** [reply]
I think you can create a custom JsonConverter that will remove the collection part from the document Id on deserialization, and then [customize the JsonDeserializer](https://ravendb.net/docs/article-page/5.1/csharp/client-api/configuration/serialization#customizejsondeserializer) of the store with that converter. But ideally its better to switch to string ids on the application itself.

**Reply by RecoX** [reply]
>ideally its better to switch to string ids on the application itself.

I am ending up doing this, as you mention this can mitigate future errors and avoid doing hacky things, still, I do not understand the use of OnAfterConversionEntity

**Comment by ayende** [comment, marked as answer]
The signature of `OnAfterConversionEntity` is: `EventHandler<AfterConversionToEntityEventArgs>` and `AfterConversionToEntityEventArgs` has read only properties for `Id`, `Entity`, and `Document`.

You are expected to transform the value to the entity.

```
 store.OnAfterConversionToEntity += (sender, args) => {
         
  if(args.Entity is EntityWithIntegerId i)
  {
        i.Id = ConvertRavenIdToNumber(args.Id);
  }
}
```

You are meant to mutate the _`Entity`_ instance
